distsqlrun: close orderedSynchronizer sources that were neglected

asubiotto opened the pull request:
These sources were neglected because they were not added to the heap
when a heap initialization error happened. The ordered synchronizer
would then go into a draining phase, but it would only close sources
that were added to the heap.

Fixes #19951

This commit also improves the testing of distsql query cancellation. The
issue that this commit fixes is caught by stress testing the new test.

Another oddity was uncovered by this: an ORDER BY DESC run after a SPLIT AT results in an ordering issue. I'll file an issue tomorrow morning and verify the minimum steps to reproduce.

I will also add orderedSynchronizer unit tests probably in another PR. Should this be cherrypicked into 1.1? I think so.

Release note (bug fix): fix a race condition that would result in some queries hanging after cancellation

Commit:
distsqlrun: close orderedSynchronizer sources that were neglected

These sources were neglected because they were not added to the heap
when a heap initialization error happened. The ordered synchronizer
would then go into a draining phase, but it would only close sources
that were added to the heap.

Fixes #19951

This commit also improves the testing of distsql query cancellation. The
issue that this commit fixes is caught by stress testing the new test.

Release note (bug fix): Fixes a race condition that would result in some
queries hanging after cancellation.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20088)
<!-- Reviewable:end -->

andreimatei commented:
don't forget release notes

I've don't have much of an idea what's going on here. If there's an error while initializing the heap, doesn't the state stay `notInitialized`, in which case we drain all the sources? Sorry, I tried reading your discussion on the bug but it's long and I lost the point :)
Let's talk in person.

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsqlrun/inbound.go, line 40 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-Kz4rqkQZGTV_LKc0fik:-Kz4rqkQZGTV_LKc0fil:b2i7m3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dba327104472330893f8fdcd61accf0056f0655/pkg/sql/distsqlrun/inbound.go#L40)):*
> ```Go
> 	// err, if set, will also be propagated to the producer
> 	// as the last record that the producer gets.
> 	if err := processInboundStreamHelper(ctx, stream, firstMsg, dst, f); err != nil {
> ```

nit: spurious change :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-:-Kz4rhZry-7ZXA2Ba4Nm:b-fhtx3y)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/input_sync.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-Kz5we3MV-WGst1OjMu1:-Kz5we3MV-WGst1OjMu2:bsfjxak) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dba327104472330893f8fdcd61accf0056f0655/pkg/sql/distsqlrun/input_sync.go#L270)):*
> ```Go
> 		// Next() calls will return metadata records.
> 		s.drainSources()
> 		// Call ConsumerClosed to primarily take care of sources that were not
> ```

This suggests that some sources may get closed some other way. Is it guaranteed that we won't call ConsumerClosed on them twice?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/distsqlrun/input_sync.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-Kz5we3MV-WGst1OjMu1:-Kz5xv_9FLpV5IqTZfHf:byi7a4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dba327104472330893f8fdcd61accf0056f0655/pkg/sql/distsqlrun/input_sync.go#L270)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This suggests that some sources may get closed some other way. Is it guaranteed that we won't call ConsumerClosed on them twice?
</blockquote></details>

In the bug you say "add a call to ConsumerClosed after the usual draining code. This will be a noop on streams that have already had ConsumerClosed called on them and will correctly close uninitialized streams" but the RowChannel doesn't tolerate that: https://github.com/cockroachdb/cockroach/blob/fc8a0ce3e4f4d716e915218a5c81c1a2f48a37b0/pkg/sql/distsqlrun/base.go#L381

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
TFTR, added release notes.

The state does not stay `notInitialized`, because the `ConsumerDone` call in the if statement handling a heap initialization error switches the state to `draining`.

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/inbound.go, line 40 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-Kz4rqkQZGTV_LKc0fik:-KzyBUhYOhuVmmn5_fh4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dba327104472330893f8fdcd61accf0056f0655/pkg/sql/distsqlrun/inbound.go#L40)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: spurious change :)
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/input_sync.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-Kz5we3MV-WGst1OjMu1:-KzyEX3WrnQoKE5NfhsY:be8a75e) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dba327104472330893f8fdcd61accf0056f0655/pkg/sql/distsqlrun/input_sync.go#L270)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

In the bug you say "add a call to ConsumerClosed after the usual draining code. This will be a noop on streams that have already had ConsumerClosed called on them and will correctly close uninitialized streams" but the RowChannel doesn't tolerate that: https://github.com/cockroachdb/cockroach/blob/fc8a0ce3e4f4d716e915218a5c81c1a2f48a37b0/pkg/sql/distsqlrun/base.go#L381
</blockquote></details>

Sorry, I'm not quite sure what I meant with what I wrote in the bug. You're right that calling `ConsumerClosed` twice will result in a panic, however it is guaranteed that we will not call `ConsumerClosed` on a stream twice. This is because once `ConsumerClosed` is called on the ordered synchronizer, it enters the `drainBuffered` state (I removed the state change below the `ConsumerClosed` call since it was unnecessary). Since we never transition from `drainBuffered` to `draining`, it's impossible to execute this code once `ConsumerClosed` has been called. This assumes that an ordered synchronizer's input streams do not get closed out from under it. In the case where `ConsumerClosed` is called after `Next`, it is specified in the `RowSource` interface that calling `ConsumerClosed` once `Next` returns an empty row is a noop. I added a state check in `ConsumerClosed` for this case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-:-KzyBDbcSa5EoM_0_klY:b5rnk9t)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
PTAL when you have a chance.

asubiotto commented:
Updated the PR with an alternative approach suggested by @andreimatei: Add sources to the heap even on error in `initHeap`, this will result in proper draining of these sources.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/run_control_test.go, line 245 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-bZPHG8P5gpLufc0Qu:-L-bZPHG8P5gpLufc0Qv:b95kooc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e074e232a91587a491cc1e5c5ebc40cb5459669b/pkg/sql/run_control_test.go#L245)):*
> ```Go
> 	// unhindered.
> 	go func() {
> 		sem <- struct{}{}
> ```

I think if `sem` was buffered (as semaphores implemented using this channel pattern usually are), you wouldn't need to do this in a separate goroutine (which is confusing).

---

*[pkg/sql/distsqlrun/input_sync.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-bX9lksOkzx9iLR5u6:-L-bX9llGqgByN-wbesf:b-xihwz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e074e232a91587a491cc1e5c5ebc40cb5459669b/pkg/sql/distsqlrun/input_sync.go#L136)):*
> ```Go
> // initHeap grabs a row from each source and initializes the heap. Any given
> // source will be on the heap (even if an error was encountered while reading
> // from it) unless there are no more rows to read from it.
> ```

please add a comment saying that if an error is returned, `heap.Init()` has not been called, so `heap` is not actually a heap. And also spell out that if an error is returned, all the members of the heap need to be drained.

---

*[pkg/sql/distsqlrun/input_sync.go, line 309 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-bY3Gy61kEEf2oMkfx:-L-bY3Gy61kEEf2oMkfy:be29p0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e074e232a91587a491cc1e5c5ebc40cb5459669b/pkg/sql/distsqlrun/input_sync.go#L309)):*
> ```Go
> // ConsumerClosed is part of the RowSource interface.
> func (s *orderedSynchronizer) ConsumerClosed() {
> 	// The state should matter, as no further methods should be called, but we'll
> ```

not your code, but I think this comment wants to say "_shouldn't_ matter"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-:-L-bZfdR60jLtVgjuHS2:bnfp4nl)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/run_control_test.go, line 245 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-bZPHG8P5gpLufc0Qu:-L-bmdDBP5mOCHX_oZK4:b-yh16l6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e074e232a91587a491cc1e5c5ebc40cb5459669b/pkg/sql/run_control_test.go#L245)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think if `sem` was buffered (as semaphores implemented using this channel pattern usually are), you wouldn't need to do this in a separate goroutine (which is confusing).
</blockquote></details>

Done. Good call.

---

*[pkg/sql/distsqlrun/input_sync.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-bX9lksOkzx9iLR5u6:-L-bmecyvJqKtAiM_-2p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e074e232a91587a491cc1e5c5ebc40cb5459669b/pkg/sql/distsqlrun/input_sync.go#L136)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment saying that if an error is returned, `heap.Init()` has not been called, so `heap` is not actually a heap. And also spell out that if an error is returned, all the members of the heap need to be drained.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/input_sync.go, line 309 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-bY3Gy61kEEf2oMkfx:-L-bmfX28zkreTZVnsNb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e074e232a91587a491cc1e5c5ebc40cb5459669b/pkg/sql/distsqlrun/input_sync.go#L309)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

not your code, but I think this comment wants to say "_shouldn't_ matter"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/input_sync.go, line 148 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-cEYc_5wsPv7lF11-S:-L-cEYc_5wsPv7lF11-T:b-rlm7ob) ([raw file](https://github.com/cockroachdb/cockroach/blob/381b57cae3a832bf4f3b85cb779cfdf8877eb61a/pkg/sql/distsqlrun/input_sync.go#L148)):*
> ```Go
> 			consumeErr = err
> 		}
> 		if src.row != nil || err != nil {
> ```

was this supposed to be &&? (in which case it could be an `else if src.row != nil`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/input_sync.go, line 148 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-cEYc_5wsPv7lF11-S:-L-cPbuNA-3WW2VoNbMX:blk7tze) ([raw file](https://github.com/cockroachdb/cockroach/blob/381b57cae3a832bf4f3b85cb779cfdf8877eb61a/pkg/sql/distsqlrun/input_sync.go#L148)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

was this supposed to be &&? (in which case it could be an `else if src.row != nil`
</blockquote></details>

This is supposed to be `||` as I want to add to the heap either if the src has a row available or there was en error while consuming its metadata.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/input_sync.go, line 148 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-cEYc_5wsPv7lF11-S:-L-cQVgI74qRr-QHpUa2:bb33esj) ([raw file](https://github.com/cockroachdb/cockroach/blob/381b57cae3a832bf4f3b85cb779cfdf8877eb61a/pkg/sql/distsqlrun/input_sync.go#L148)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This is supposed to be `||` as I want to add to the heap either if the src has a row available or there was en error while consuming its metadata.
</blockquote></details>

Oh. What's going on here is pretty subtle and needs more comments. Explain that we're still adding the source just so it's in `s.heap` for cleanup purposes (see `consumerStatusChanged`), we won't actually initialize the heap with a nil row.

In fact, I'm still missing something.. `initHeap` is only called in `notInitialized` state and if it returns an error, we will remain in that state and `consumerStatusChanged` won't look at the heap.


Instead of this, maybe we should add a `needsClosing util.FastIntSet` with the set of sources that weren't closed. `consumerStatusChanged` would just consult this set instead of behaving differently based on `state`. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/distsqlrun/input_sync.go, line 148 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-L-cEYc_5wsPv7lF11-S:-L-h0H0Wu8K0XiWTHXmz:b-7s99gw) ([raw file](https://github.com/cockroachdb/cockroach/blob/381b57cae3a832bf4f3b85cb779cfdf8877eb61a/pkg/sql/distsqlrun/input_sync.go#L148)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Oh. What's going on here is pretty subtle and needs more comments. Explain that we're still adding the source just so it's in `s.heap` for cleanup purposes (see `consumerStatusChanged`), we won't actually initialize the heap with a nil row.

In fact, I'm still missing something.. `initHeap` is only called in `notInitialized` state and if it returns an error, we will remain in that state and `consumerStatusChanged` won't look at the heap.


Instead of this, maybe we should add a `needsClosing util.FastIntSet` with the set of sources that weren't closed. `consumerStatusChanged` would just consult this set instead of behaving differently based on `state`. What do you think?
</blockquote></details>

There are two `Next()` calls to consider:

1) The first calls `initHeap` and gets back an error, running `ConsumerDone()`, as you rightly pointed out, on all the sources. This call to `consumerStatusChanged` also changes the synchronizer's state to `draining`. The error is returned through metadata, so the downstream component will call Next() again.
2) The second call will then proceed to the `draining` conditional in `Next()`, this is the work that actually drains the sources in `drainSources`. The `drainSources` method only touches the heap though, so does not consume any data from sources not in the heap.

The `needsClosing` idea is good but I think is effectively implemented with this change. I don't want to add any additional state in this PR.

Added a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20088#-:-L-h6HKZSUO4RIl_iJwe:bnfp4nl)*
<!-- Sent from Reviewable.io -->
